File: tools/json_comments.py
from __future__ import annotations

import json
from typing import Any


def loads_json_with_comments(source: str) -> Any:
    return json.loads(strip_json_comments(source))


def strip_json_comments(source: str) -> str:
    output: list[str] = []
    in_string = False
    escaped = False
    in_line_comment = False
    in_block_comment = False
    index = 0

    while index < len(source):
        char = source[index]
        next_char = source[index + 1] if index + 1 < len(source) else ""

        if in_line_comment:
            if char in "\r\n":
                in_line_comment = False
                output.append(char)
            else:
                output.append(" ")
            index += 1
            continue

        if in_block_comment:
            if char == "*" and next_char == "/":
                output.append("  ")
                index += 2
                in_block_comment = False
            else:
                output.append(char if char in "\r\n" else " ")
                index += 1
            continue

        if in_string:
            output.append(char)
            if escaped:
                escaped = False
            elif char == "\\":
                escaped = True
            elif char == '"':
                in_string = False
            index += 1
            continue

        if char == '"':
            in_string = True
            output.append(char)
        elif char == "/" and next_char == "/":
            output.append("  ")
            index += 1
            in_line_comment = True
        elif char == "/" and next_char == "*":
            output.append("  ")
            index += 1
            in_block_comment = True
        else:
            output.append(char)
        index += 1

    return "".join(output)

File: exp-platform/rust/shared/src/json_comments.rs
pub fn strip_json_comments(source: &str) -> String {
    let chars: Vec<char> = source.chars().collect();
    let mut output = String::with_capacity(source.len());
    let mut in_string = false;
    let mut escaped = false;
    let mut in_line_comment = false;
    let mut in_block_comment = false;
    let mut index = 0;

    while index < chars.len() {
        let char = chars[index];
        let next = chars.get(index + 1).copied();

        if in_line_comment {
            if matches!(char, '\n' | '\r') {
                in_line_comment = false;
                output.push(char);
            } else {
                output.push(' ');
            }
            index += 1;
            continue;
        }

        if in_block_comment {
            if char == '*' && next == Some('/') {
                output.push_str("  ");
                index += 2;
                in_block_comment = false;
            } else {
                output.push(if matches!(char, '\n' | '\r') {
                    char
                } else {
                    ' '
                });
                index += 1;
            }
            continue;
        }

        if in_string {
            output.push(char);
            if escaped {
                escaped = false;
            } else if char == '\\' {
                escaped = true;
            } else if char == '"' {
                in_string = false;
            }
            index += 1;
            continue;
        }

        if char == '"' {
            in_string = true;
            output.push(char);
        } else if char == '/' && next == Some('/') {
            output.push_str("  ");
            index += 1;
            in_line_comment = true;
        } else if char == '/' && next == Some('*') {
            output.push_str("  ");
            index += 1;
            in_block_comment = true;
        } else {
            output.push(char);
        }
        index += 1;
    }

    output
}
